### ISSUE TITLE
Modify displayed call signature for astropy.coordinates frame classes?

### ISSUE THREAD

**Issue by adrn** [astropy/astropy#10894]
### Description

*Note: This is a question meant to raise discussion, not really a feature request. I am also not strongly advocating for this, but interested to hear thoughts!*

The frame classes in `astropy.coordinates` accept a flexible set of input arguments, and the argument names depend on arguments that can be passed in (e.g., `representation_type` determines the coordinate component names that are allowed). The call signatures (both as displayed by `help()` and as implemented in code) therefore rely heavily on `*args` and `**kwargs`. For example, calling `help(coord.ICRS)` (or `coord.ICRS(`<tab> to pull up a docstring in IPython) outputs:

```ICRS(*args, copy=True, representation_type=None, differential_type=None, **kwargs)```

Since `__init_subclass__` was added in Python 3.6, we now have had the option of modifying the call signatures automatically for each class based on the default component names and frame attributes (which are defined at the class level for each of the frame classes). So, for example, the *code* could stay exactly as it is, but we could override the *displayed* call signature in `BaseCoordinateFrame.__init_subclass__` to reflect the default set of arguments that are accepted. So, for example, instead of the above for ICRS, it could *show* something like

coord.Galactic(
    *args,
    copy=True,
    representation_type=None,
    differential_type=None,
    **kwargs,
)
```ICRS(
    ra, 
    dec, 
    distance=None, 
    pm_ra_cosdec=None, 
    pm_dec=None, 
    radial_velocity=None, 
    representation_type=None, 
    differential_type=None, 
    copy=True, 
    **kwargs,
)```

The advantage here is that right now, users have to scroll down to the "Parameters" block to know what input is accepted - this would give them a quicker idea of the allowed arguments. The disadvantage would be, as I noted above, the input is actually quite a bit more flexible and the component names *depend on* other keyword arguments. So: Would this be a good idea? I am *slightly* in favor, because I do think it's better to show the default argument names instead of just *args, but not strongly in favor...

(This was on my mind because I just added something similar to my affiliated package Gala, e.g., 
 https://github.com/adrn/gala/blob/main/gala/potential/common.py#L45)

cc @mhvk @eteq @astrojuanlu @StuartLittlefair

**Comment by mhvk** [comment]
Currently, `ICRS?` gives
```
Init signature:
ICRS(
    *args,
    copy=True,
    representation_type=None,
    differential_type=None,
    **kwargs,
)
Docstring:     
A coordinate or frame in the ICRS system.

If you're looking for "J2000" coordinates, and aren't sure if you want to
use this or `~astropy.coordinates.FK5`, you probably want to use ICRS. It's
more well-defined as a catalog coordinate and is an inertial system, and is
very close (within tens of milliarcseconds) to J2000 equatorial.

For more background on the ICRS and related coordinate transformations, see
the references provided in the  :ref:`astropy-coordinates-seealso` section
of the documentation.

Parameters
----------
data : `~astropy.coordinates.BaseRepresentation` subclass instance
    A representation object or ``None`` to have no data (or use the
    coordinate component arguments, see below).

ra : `~astropy.coordinates.Angle`, optional, must be keyword
    The RA for this object (``dec`` must also be given and ``representation``
    must be None).
dec : `~astropy.coordinates.Angle`, optional, must be keyword
    The Declination for this object (``ra`` must also be given and
    ``representation`` must be None).
distance : `~astropy.units.Quantity`, optional, must be keyword
    The Distance for this object along the line-of-sight.
    (``representation`` must be None).
```

I think the main thing making this less clear is having `data` first - I think that is a leftover (arguably we should just have only keyword arguments). But otherwise it seems not far down?

**Comment by mhvk** [comment]
Actually, shouldn't we look mostly at making `SkyCoord` as clear as possible? I thought the interface via frames was not really the recommended one for all but experts?

I guess overall this seems a bit more magic than is warranted (though I like `__init_subclass__` and at the very least there places where we use a metaclass now that could benefit from it)

**Comment by adrn** [comment]
I meant call signature or "Init signature" as it's listed above, not the whole docstring (but there could be a separate issue for auditing those!). Yes, it's a fair point that there is more info under the `Parameters` heading, but I often try to infer a lot from the init signature alone when coding because that is what Jupyter/IPython shows when hitting `tab` after an open parenthesis for a function or class init call:
<img width="761" alt="Screen Shot 2020-10-18 at 9 29 26 PM" src="https://user-images.githubusercontent.com/583379/96392159-327a4e00-1189-11eb-9482-ac03cd6316cd.png">

**Comment by mhvk** [comment]
Ah, yes, now I see where you are coming from... that is decidedly unhelpful. And indeed I often use tab-completion to get keyword arguments, which wouldn't work here either.

**Comment by ayshih** [comment]
> The disadvantage would be, as I noted above, the input is actually quite a bit more flexible and the component names _depend on_ other keyword arguments.

This aspect makes me nervous, because it might lead users to think, for example, that there's no exposed interface for defining an `ICRS` coordinate using Cartesian components, given that those keywords are not shown.  They might believe they have to manually convert their components to the "correct" representation prior to input.  Of course, the reason that users would be less likely to have that confusion under the current presentation is because the signature is totally unhelpful...

As @mhvk noted, `SkyCoord` is what users should be using, and I presume some fancier magic would be needed to get something like this to work there.

**Comment by eteq** [comment]
Hmm.  I tilt toward having the same concern as @ayshih - in @adrn's initial example it really does seem like the only valid inputs are (ra, dec) which isn't quite right here.  And I'm also sympathetic to the idea that the frame classes are meant for someone with a layer deeper of understanding and thus it's really `SkyCoord` to focus on for this stuff.

That said... I sympathize with the underlying concern, particularly when I have do the "wait is it `pm_cosdec_ra`? or `pm_ra_cosdec`? or `pm_cosra_dec`? is up down?  Anything could be true, it's 2020!" dance...

Having not explored `__init_subclass__` much myself, maybe @adrn has insight: is it possible to end up with a signature that looks something like:
```
ICRS(
    *,
    ra?, 
    dec?, 
    distance=None, 
    pm_ra_cosdec=None, 
    pm_dec=None, 
    radial_velocity=None, 
    representation_type=None, 
    differential_type=None, 
    copy=True, 
    **kwargs,
)
```
? That sort of thing would provide the context hint for those who know while still forcing those who aren't sure what's happening to read down to somewhere that will tell them what that ``*`` means...

**Comment by astrojuanlu** [comment]
I agree with the concern @ayshih raised, although I also find myself messing up input parameters and would like to have a more helpful docstring. Don't see myself using SkyCoord though, base frames serve me well in poliastro and I don't want a higher level interface.

Perhaps this is only tangentially related, but I wonder if we could move towards offering classmethods with a simpler API. For example:

```
class ICRS:
    @classmethod
    def from_representation(cls, data, **kwargs): ...
    @classmethod
    def from_spherical(cls, ra, dec, **kwargs): ...
```

I did not evaluate all the implications of this (hopefully we wouldn't have to do it in every frame subclass). Even if they are kept private at first, it would simplify the `__init__` code a lot (which is a pain point at least for me, see my so far unsuccessful #10788)

**Comment by adrn** [comment]
RE: `SkyCoord`, I said "frame classes" in the issue, but I think the same question stands for modifying the `SkyCoord` init signature. I guess the main question is: Is it better to have an unhelpful init signature that forces users to look at the docstring to see how to initialize, or have an init signature that is helpful for some large fraction of users but could lead to confusion for users that want to be able to pass in more flexible input? I'm about 50/50 on that draw...

@astrojuanlu RE: classmethods. Yes, that is the dream! I think I and/or @eteq and/or @mhvk started down that path in some branch or PR and got stuck on the fact that the `SkyCoord` initializer is truly a beast. We have made some simplifications to it over the years, but fully re-implementing it in a simpler way that uses classmethods and still passes all previous tests is a big project that I think none of us have had the time or space for :/

**Comment by astrojuanlu** [comment]
I have been thinking about these classmethods a bit more. The sad thing is that every frame would need to implement a different `from_{rep}` method, according to their `frame_specific_representation_info`: for example, `ICRS` would have `from_spherical(self, ra, dec, ...)` whereas `*Ecliptic` would have `from_spherical(self, lon, lat, ...)`.

...or we could accept this duplication, but perhaps it would also be a maintainability burden.

...or they could have a generic `*args` signature, but then the objective of making them more explicit is not achieved.

...or we could magically generate these methods from the `frame_specific_representation_info`, but I'm not sure I'd want to add more magic...
